### DISCUSSION TITLE
Using System.Data.SqlServer in 3.1.0

### DISCUSSION THREAD

**Discussion by asherber** [quartznet/quartznet#950] [General]
I know that v3.1.0 uses Microsoft.Data.SqlClient by default; I'm trying to get it to use System.Data.SqlClient instead. I've copied the appropriate lines from `dbproviders.properties` and put them into a `quartz.config` file in my application's root, but my app seems to still be expecting Microsoft.Data.SqlClient.

I'm using a custom `DbProvider` descendant which overrides `CreateConnection()` to explicitly hand back a `System.Data.SqlConnection` from some middleware. When I instantiate this provider, I can see right away that the `Metadata` is for Microsoft.Data, meaning that `DbProvider` doesn't seem to be reading `quartz.config'. And when Quartz tries to access the db, the error message I get is "Unable to cast object of type 'System.Data.SqlClient.SqlConnection' to type 'Microsoft.Data.SqlClient.SqlConnection'."

Am I missing a step here?

**Comment by lahma** [comment]
Yes this thing backfired a bit with 3.1 release. The upcoming 3.2 release will revert back to using System.Data.SqlConnection again so I'd suggest that you point your NuGet to [MyGet feed](https://www.myget.org/F/quartznet/api/v3/index.json) which contains the latest fixes from master. 

[Here's the PR for reference](https://github.com/quartznet/quartznet/pull/921/files).

**Reply by devigo** [reply]
Can we apply almost the same fix for .NET Core? For example https://github.com/devigo/quartznet/compare/main...devigo:system-data-sqlclient-netstandard-tweaks?expand=1
If we want to use `System.Data.SqlServer` in .NET Core we can specify it explicitly with this parameter
`quartz.dataSource.myDS.provider = SystemDataSqlClient`

**Comment by asherber** [comment]
Thanks, I saw the PR, and I know I could use the new builds or revert to 3.0.7. But is there no way to tell 3.1.0 to use System.Data.Sql?

**Reply by lahma** [reply]
Make sure you use the correct format for registering, [here's an example](https://github.com/quartznet/quartznet/blob/139aafa23728892b0a5ebf845ce28c3bfdb0bfe8/src/Quartz.Examples/quartz.config) which register custom provider with name `customProvider`.

**Reply by asherber** [reply]
I think `DbProvider` is not correctly reading `quartz.config` in the constructor. My file has:

```
quartz.dbprovider.SqlServerCustom.productName=Microsoft SQL Server
quartz.dbprovider.SqlServerCustom.assemblyName=System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
etc.
```

But when I do `new DbProvider("SqlServerCustom", myConnString)`, I get "There is no metadata information for provider 'SqlServerCustom'"

I have verified that `quartz.config` is correctly copied to the output directory.

**Reply by lahma** [reply]
You could also try to do things programmatically via `DbProvider.RegisterDbMetadata`

**Reply by asherber** [reply]
Right, but `DbProvider` is supposed to read `quartz.config`, isn't it? I've got a minimal project which reproduces this issue, if you want.

**Reply by asherber** [reply]
So it looks like the `ConfigurationBasedDbMetadataFactory` which `DbProvider` uses only reads `app.config` (via `ConfigurationManager`), not the `quartz.config` properties file. Is this by design?

**Reply by lahma** [reply]
Sounds like a bug as they should be quite interchangeable.

**Reply by lahma** [reply]
Quartz 3.2 has been released with the fixes.

**Comment by devigo** [comment]
We had the same issue in .NET Core 2.1. To fix this, you can put the following content into a` quartz.config`
```
# alias for System.Data.SqlClient (from dbproviders.properties)
quartz.dbprovider.SystemDataSqlClient.productName=Microsoft SQL Server
quartz.dbprovider.SystemDataSqlClient.assemblyName=System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
quartz.dbprovider.SystemDataSqlClient.connectionType=System.Data.SqlClient.SqlConnection, System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
quartz.dbprovider.SystemDataSqlClient.commandType=System.Data.SqlClient.SqlCommand, System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
quartz.dbprovider.SystemDataSqlClient.parameterType=System.Data.SqlClient.SqlParameter, System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
quartz.dbprovider.SystemDataSqlClient.parameterDbType=System.Data.SqlDbType, System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
quartz.dbprovider.SystemDataSqlClient.parameterDbTypePropertyName=SqlDbType
quartz.dbprovider.SystemDataSqlClient.parameterNamePrefix=@
quartz.dbprovider.SystemDataSqlClient.exceptionType=System.Data.SqlClient.SqlException, System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
quartz.dbprovider.SystemDataSqlClient.useParameterNamePrefixInParameterCollection=true
quartz.dbprovider.SystemDataSqlClient.bindByName=true
quartz.dbprovider.SystemDataSqlClient.dbBinaryTypeName=VarBinary
```
or
```
# alias for System.Data.SqlClient (from dbproviders.netstandard.properties)
quartz.dbprovider.SystemDataSqlClient.productName=Microsoft SQL Server Core
quartz.dbprovider.SystemDataSqlClient.assemblyName=System.Data, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
quartz.dbprovider.SystemDataSqlClient.connectionType=System.Data.SqlClient.SqlConnection, System.Data.SqlClient, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
quartz.dbprovider.SystemDataSqlClient.commandType=System.Data.SqlClient.SqlCommand, System.Data.SqlClient, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
quartz.dbprovider.SystemDataSqlClient.parameterType=System.Data.SqlClient.SqlParameter, System.Data.SqlClient, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
quartz.dbprovider.SystemDataSqlClient.parameterDbType=System.Data.SqlDbType, System.Data.SqlClient, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
quartz.dbprovider.SystemDataSqlClient.parameterDbTypePropertyName=SqlDbType
quartz.dbprovider.SystemDataSqlClient.parameterNamePrefix=@
quartz.dbprovider.SystemDataSqlClient.exceptionType=System.Data.SqlClient.SqlException, System.Data.SqlClient, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
quartz.dbprovider.SystemDataSqlClient.useParameterNamePrefixInParameterCollection=true
quartz.dbprovider.SystemDataSqlClient.bindByName=true
quartz.dbprovider.SystemDataSqlClient.dbBinaryTypeName=VarBinary
```
Both work for us in Quartz 3.2.4
Note that we use a custom name for the provider (`SystemDataSqlClient`) instead of `SqlServer`.
Also don't forget to update the provider name in your configuration, like this
![image](https://user-images.githubusercontent.com/34477515/109429517-1fb43780-7a0d-11eb-8fef-fb0ee792c6e4.png)
